class_13_parallel_programming/homework/homework.py

```diff
@@ -0,0 +1,291 @@
+import argparse
+import io
+import logging
+import os
+import sys
+import time
+from contextlib import contextmanager
+from functools import wraps
+from multiprocessing import Lock
+from multiprocessing.pool import ThreadPool
+from pathlib import Path, PurePath
+from typing import Any, Iterable, List, Tuple
+
+import requests
+from PIL import Image
+
+# not sure how to set it up correctly. Used by `handle_image_processing` function
+ProcessParams = Tuple[str, Tuple[int, int], int, Lock, Iterable[Tuple[int, int]]]
+
+# set up logger
+LOG_FORMAT = "%(levelname)s: '%(message)s'"
+logging.basicConfig(format=LOG_FORMAT, level=logging.INFO, stream=sys.stdout)
+
+
+# utility functions
+@contextmanager
+def move_to_dir(path: Path) -> None:
+    """Context manager for getting into a provided directory"""
+
+    tmp_path = os.getcwd()
+    os.chdir(path)
```

Comment:
Is not it supposed to be included in try section?

Reply:
I hope it is not. In my understanding we can change cwd to be iput parameter `path` (which was already validated by this time). Example is taken from [here](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager).
So my thought is that if we don't need to yield directory we move in and don't need to validate it, we can leave this code outside `try` block.
Please let me know if I'm mistaken about it.

Reply:
But in this case why do you need try block? Do you think logging.debug or yield can throw an exception?

Reply:
If I understand documentation for `contexmanager` decorator correctly I have to put `yield` inside `try - finally` to be able to do a cleanup (return to original directory) in case there was an unhandled exception raised inside `with move_to_dir` block.
this is how I interpreted the following from the link above:

> At the point where the generator yields, the block nested in the with statement is executed. The generator is then resumed after the block is exited. If an unhandled exception occurs in the block, it is reraised inside the generator at the point where the yield occurred. Thus, you can use a try…except…finally statement to trap the error (if any), or ensure that some cleanup takes place.

logging could go before `try` though so only `yield` will be inside `try`, I will fix it.